library/psa_crypto.c

```diff
@@ -3799,7 +3799,7 @@ psa_status_t psa_verify_hash( psa_key_handle_t handle,
 #if defined(MBEDTLS_ECP_C)
     if( PSA_KEY_TYPE_IS_ECC( slot->attr.type ) )
     {
-#if defined(MBEDTLS_ECDSA_C)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_ECDSA)
```

Comment:
```suggestion
#if defined(MBEDTLS_PSA_BUILTIN_ALG_ECDSA) || defined(MBEDTLS_PSA_BUILTIN_ALG_DETERMINISTIC_ECDSA)
```
For verification, ECDSA and deterministic ECDSA are identical.

Also applies at line 3820 below.

Reply:
When I attempted this change before, it broke several of the CI tests. I think there is more work to be done to properly address that and should probably be done in a separate ticket. We are essentially changing this definition from only being enabled when MBEDTLS_ECDSA_C to when that or MBEDTLS_ECDSA_DETERMINISTIC which I think is going beyond what this ticket is trying to accomplish. It will be harder to untangle the changes to simply move to this new guarding to also fixing other dependencies when using this new guarding. If you do not agree to keep this a separate issue, I will fix and start trying to figure out the dependency issues with the CI tests when that is modified.

Reply:
@gilles-peskine-arm Forgot to tag your name on my previous comment. It is the last item to address in this round of review that I did not make changes for yet. Let me know how I should proceed based on my comment.

Reply:
What broke? In Mbed TLS as it is now, `MBEDTLS_PSA_BUILTIN_ALG_DETERMINISTIC_ECDSA` implies `MBEDTLS_ECDSA_C` which is sufficient for `MBEDTLS_PSA_BUILTIN_ALG_ECDSA`, so I don't see how making this test correct has a material impact: I requested this for robustness against future evolution.

Reply:
@gilles-peskine-arm It was back when you first made the comment to make the change. Once I completed that change it resulted in several CI build failures. Here is your comment from earlier.

"As of a8b01e1 the following all.sh components are failing: test_everest_curve25519_only test_depends_curves test_depends_curves_psa test_depends_pkalgs test_depends_pkalgs_psa. All the failures seem to be when ECDSA is disabled, with some code in psa_crypto.c trying to call ECDSA functions nonetheless."

I have a feeling that there is likely a test where ECDSA_C is being disabled, yet ECDSA_DETERMINISTIC is set and because the requirement to include ECDSA_C if ECDSA_DETERMINISTIC is not checked when adding this code in it can cause failures. I agree with you that this change needs to be made, I just think it would make the change set more clear if it is not cluttered with all these other changes and given it's own ticket to have the changes tied to.

Reply:
If you don't see an issue with making these changes now, I'll dig further and see if I can't get them fixed quickly and add them to this PR.

Reply:
Is this still an issue now that [`PSA_WANT_ALG_DETERMINISTIC_ECDSA` enables `MBEDTLS_ECDSA_C` if there's no accelerator](https://github.com/ARMmbed/mbedtls/pull/3712#discussion_r512018836)?

Reply:
Yes, I just tried making the changes again and running test_everest_curve25519_only and had these errors.

psa_crypto.c: In function ‘psa_ecdsa_sign’:
psa_crypto.c:3563:26: error: implicit declaration of function ‘mbedtls_ecdsa_sign_det_ext’ [-Werror=implicit-function-declaration]
         MBEDTLS_MPI_CHK( mbedtls_ecdsa_sign_det_ext( &ecp->grp, &r, &s,
                          ^
../include/mbedtls/bignum.h:50:22: note: in definition of macro ‘MBEDTLS_MPI_CHK’
         if( ( ret = (f) ) != 0 ) \
                      ^
psa_crypto.c:3573:26: error: implicit declaration of function ‘mbedtls_ecdsa_sign’ [-Werror=implicit-function-declaration]
         MBEDTLS_MPI_CHK( mbedtls_ecdsa_sign( &ecp->grp, &r, &s, &ecp->d,
                          ^
../include/mbedtls/bignum.h:50:22: note: in definition of macro ‘MBEDTLS_MPI_CHK’
         if( ( ret = (f) ) != 0 ) \
                      ^
psa_crypto.c: In function ‘psa_ecdsa_verify’:
psa_crypto.c:3624:11: error: implicit declaration of function ‘mbedtls_ecdsa_verify’ [-Werror=implicit-function-declaration]
     ret = mbedtls_ecdsa_verify( &ecp->grp, hash, hash_length,
           ^
psa_crypto.c: At top level:
psa_crypto.c:3537:21: error: ‘psa_ecdsa_sign’ defined but not used [-Werror=unused-function]
 static psa_status_t psa_ecdsa_sign( mbedtls_ecp_keypair *ecp,

Reply:
Then something else is wrong, because neither `MBEDTLS_PSA_BUILTIN_ALG_ECDSA` nor `MBEDTLS_PSA_BUILTIN_ALG_DETERMINISTIC_ECDSA` should be enabled in that configuration.

Reply:
@gilles-peskine-arm It has to be setting one of those items because if I do not have the changes being requested, then all those tests are passing. I'll see if I can root cause exactly what is happening.

Reply:
Looks like for that particular test the config.h has MBEDTLS_ECDSA_C disabled, but MBEDTLS_DETERMINISTIC_ECDSA is set which explains why the test has the errors. Now, if MBEDTLS_DETERMINISTIC_ECDSA should not be set for that test, then I guess the test needs to be updated to fix that issue.

Reply:
With `MBEDTLS_PSA_CRYPTO_CONFIG` disabled, if `MBEDTLS_ECDSA_C` is disabled, the state of `MBEDTLS_DETERMINISTIC_ECDSA` doesn't change anything. Your PR shouldn't break that (if it's breaking one of our tests, there's a good chance that it also breaks someone's production code).

Reply:
@gilles-peskine-arm So the way the code works now, if MBEDTLS_PSA_CRYPTO_CONFIG is disabled, the MBEDTLS_ECDSA_C sets the MBEDTLS_PSA_BUILTIN_ALG_ECDSA and if MBEDTLS_ECDSA_DETERMINISTIC is set, then MBEDTLS_PSA_BUILTIN_ALG_DETERMINISTIC_ECDSA is set. The new change requested for this block is essentially to change a guard around a block of code that was only enabled when ECDSA_C was set. That guard is now changed to be enabled when ECDSA_C is set or DETERMINISTIC is set. Even if CRYPTO_CONFIG is disabled the guards in the library code have changed so test results could also change if those tests are not properly ensuring the setting of those two macros. I can update the tests to ensure if a test disables ECDSA_C it also disables DETERMINISTIC which is essentially needed if we update the guards. Looking at the tests, there seem to be several that to disable ECDSA_C they only disable these macros.

unset MBEDTLS_ECDSA_C
unset MBEDTLS_KEY_EXCHANGE_ECDH_ECDSA_ENABLED
unset MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED

Is the correct way to fix this by updating the tests to also ensure the DETERMINISTIC is also disabled, or is there a different approach in config_psa.h where when CRYPTO_CONFIG is disabled we automatically handle this?

Reply:
There are two potential sensible meanings for `MBEDTLS_ECDSA_DETERMINISTIC`:

1. If enabled, deterministic ECDSA is supported, otherwise deterministic ECDSA is not supported. In this approach, enabling `MBEDTLS_ECDSA_DETERMINISTIC` without `MBEDTLS_ECDSA_C` is a configuration error.
2. If enabled, the ECDSA module supports deterministic ECDSA. If disabled, the ECDSA module does not support deterministic ECDSA. In this approach, the status of `MBEDTLS_ECDSA_DETERMINISTIC` is indifferent if `MBEDTLS_ECDSA_C` is disabled.

Mbed TLS currently follows approach 2. This may not have been the right choice, but that's what it is. I don't think this pull request is the time or place to change it.

In this pull request, in `config_psa.h`, if `MBEDTLS_PSA_CRYPTO_CONFIG` is disabled, `MBEDTLS_ECDSA_DETERMINISTIC` should only be taken into account if `MBEDTLS_ECDSA_C` is enabled.

Reply:
Ok, thanks for the clarification. I'll make the updates to config_psa.h and the library code and see if it passes all the tests.